examples/run_summarization_finetuning.py

```diff
@@ -0,0 +1,625 @@
+        return len(self.source_examples)
+
+    def __getitem__(self, items):
+        return (
+            torch.tensor(self.source_examples[items]),
+            torch.tensor(self.target_examples[items]),
+        )
+
+
+def process_story(raw_story, tokenizer):
+    """ Extract the story and summary from a story file.
+
+    Attributes:
+        raw_story (str): content of the story file as an utf-8 encoded string.
+
+    Raises:
+        IndexError: If the stoy is empty or contains no highlights.
+    """
+    file_lines = list(
+        filter(lambda x: len(x) != 0, [line.strip() for line in raw_story.split("\n")])
+    )
+
+    # for some unknown reason some lines miss a period, add it
+    file_lines = [_add_missing_period(line) for line in file_lines]
+
+    # gather article lines
+    story_lines = []
+    lines = deque(file_lines)
+    while True:
+        try:
+            element = lines.popleft()
+            if element.startswith("@highlight"):
+                break
+            story_lines.append(element)
+        except IndexError as ie:  # if "@highlight" absent from file
+            raise ie
+
+    # gather summary lines
+    highlights_lines = list(filter(lambda t: not t.startswith("@highlight"), lines))
+
+    # join the lines. This sequence is specific to Lapata & Liu (2019).
+    # Edit when you are adding a new model.
+    story_ids = [
+        tokenizer.add_special_tokens_single_sequence(tokenizer.encode(story))
+        for story in story_lines
+    ]
+    summary_ids = [
+        tokenizer.add_special_tokens_single_sequence(tokenizer.encode(highlight))
+        for highlight in highlights_lines
+    ]
+
+    flattened_story_ids = [token for sentence in story_ids for token in sentence]
+    flattened_summary_ids = [token for sentence in summary_ids for token in sentence]
+
+    return flattened_story_ids, flattened_summary_ids
+
+
+def _add_missing_period(line):
+    END_TOKENS = [".", "!", "?", "...", "'", "`", '"', u"\u2019", u"\u2019", ")"]
+    if line.startswith("@highlight"):
+        return line
+    if line[-1] in END_TOKENS:
+        return line
+    return line + "."
+
+
+def _fit_to_block_size(sequence, block_size):
+    """ Adapt the source and target sequences' lengths to the block size.
+    If the sequence is shorter than the block size we pad it with -1 ids
+    which correspond to padding tokens.
+    """
+    if len(sequence) > block_size:
+        return sequence[:block_size]
+    else:
+        sequence.extend([0] * (block_size - len(sequence)))
+        return sequence
+
+
+def mask_padding_tokens(sequence):
+    """ Replace the padding token with -1 values """
+    padded = sequence.clone()
+    padded[padded == 0] = -1
+    return padded
+
+
+def load_and_cache_examples(args, tokenizer):
+    dataset = TextDataset(tokenizer, data_dir=args.data_dir)
+    return dataset
+
+
+def compute_token_type_ids(batch, separator_token_id):
+    """ Segment embeddings as described in [1]
+
```

Comment:
Document the input. I’d prefer a more elegant/vectorized solution but this is legible and does the job atm.